design/one-pager-easy-stacks.md

```diff
@@ -0,0 +1,294 @@
+
+## Technical Implementation
+
+### Metadata of The Stack
+
+[Dependency list] has to be filled with each kind of the resource that the stack
+deploys, i.e. f the stack only deploys `RDSInstanceClass` resource, it should
+not require all AWS stack to be present.
+
+[UI annotations] should be present to make it easy for frontend software to process
+the stack.
+
+### Folder Structure
+
+All directories containing base YAML files should be in a top-level directory
+of the repository named `resources` so that it's easier for others to fork & change.
+
+Controller will scan this folder recursively to get all YAML files. So, the folder
+structure inside that `resources` directory won't matter to the controller but
+it's good to separate the folders cleanly so that it's easier to fork & change.
+
+There are mainly 2 types of resources that configuration stacks will usually
+need to deploy; resource classes and managed resources. Under `resources` folder
+we can have 2 separate folders for each of those resource types; `classes` and
+`services`. Under these folders, resources should be in directories recursively
+named after their group and kind. An example folder structure:
+
+```
+├── resources
+│   ├── classes
+│   │   ├── cache
+│   │   │   └── replicationgroup.yaml
+│   │   ├── compute
+│   │   │   └── ekscluster.yaml
+│   │   └── database
+│   │       ├── rdsinstancemysql.yaml
+│   │       └── rdsinstancepostgresql.yaml
+│   └── services
+│       ├── database
+│       │   └── dbsubnetgroup.yaml
+│       ├── identity
+│       │   ├── iamrole.yaml
+│       │   └── iamrolepolicyattachment.yaml
+│       └── network
+│           ├── internetgateway.yaml
+│           ├── routetable.yaml
+│           ├── securitygroup.yaml
+│           ├── subnet.yaml
+│           └── vpc.yaml
+```
+
+Example above is a suggestion for stacks with one tier of configuration.
+There could also be cases where one configuration stack has the YAML files for
+different tiers and a `spec` allows to deploy one of them, in that case it'd make
+sense for developer to have a folder for each tier under `resources` folder.
+
+### Reconciliation
+
+There are mainly 3 responsibility of the controller:
+* When all deployed resources are ready, `status.state` should be `ready`.
+  * For resource classes; if the resource exists in the api-server it's considered
+    to be ready.
+  * For managed resources; if the `Ready` condition status is `true`, then it's
+    considered to be ready.
+  * For other possible resources, the developer chooses what to look for as
+    readiness.
+* Apply; continuously reproduce the resources from YAMLs and apply it.
+* Deletion: should be done via label selector when the CR is deleted.
+
+In the `Apply` phase, there will be some processing phasse before calling `Create`
+for the resources:
+
+1. Names of the resources should be converted to the following format
+  `<CR namespace>-<CR name>-<Name given in YAML>` to prevent name collisions
```

Comment:
Is there a name length limit, and could we run into problems? Another option is to include the CR UID in the name, though it's a little less human-friendly.

Reply:
I believe 63 is the character limit. Since it's now cluster-scoped, we'll only use the CR's name and I think `uid` is longer than most of the CR names. So, it should be safer to go with name.  But, in general, yes if we hapen to have very long CR name and resource name, we'll run into a problem.